src/ripple/legacy/0.27/CreateOffer.cpp

```diff
@@ -0,0 +1,631 @@
+
+        return terResult;
+    }
+
+    if (m_journal.debug)
+    {
+        m_journal.debug <<
+            "takeOffers: saTakerPays=" <<saTakerPays.getFullText ();
+        m_journal.debug <<
+            "takeOffers: saTakerGets=" << saTakerGets.getFullText ();
+        m_journal.debug <<
+            "takeOffers: mTxnAccountID=" <<
+            to_string (mTxnAccountID);
+        m_journal.debug <<
+            "takeOffers:         FUNDS=" << view.accountFunds (
+            mTxnAccountID, saTakerGets, fhZERO_IF_FROZEN).getFullText ();
+    }
+
+    if (saTakerPays < zero || saTakerGets < zero)
+    {
+        // Earlier, we verified that the amounts, as specified in the offer,
+        // were not negative. That they are now suggests that something went
+        // very wrong with offer crossing.
+        m_journal.fatal << (crossed ? "Partially consumed" : "Full") <<
+            " offer has negative component:" <<
+            " pays=" << saTakerPays.getFullText () <<
+            " gets=" << saTakerGets.getFullText ();
+
+        assert (saTakerPays >= zero);
+        assert (saTakerGets >= zero);
+        return tefINTERNAL;
+    }
+
+    if (bFillOrKill && (saTakerPays != zero || saTakerGets != zero))
+    {
+        // Fill or kill and have leftovers.
+        view.swapWith (view_checkpoint); // Restore with just fees paid.
+        return tesSUCCESS;
+    }
+
+    // What the reserve would be if this offer was placed.
+    auto const accountReserve (mEngine->getLedger ()->getReserve (
+        sleCreator->getFieldU32 (sfOwnerCount) + 1));
+
+    if (saTakerPays == zero ||                // Wants nothing more.
+        saTakerGets == zero ||                // Offering nothing more.
+        bImmediateOrCancel)                   // Do not persist.
+    {
+        // Complete as is.
+    }
+    else if (mPriorBalance.getNValue () < accountReserve)
+    {
+        // If we are here, the signing account had an insufficient reserve
+        // *prior* to our processing. We use the prior balance to simplify
+        // client writing and make the user experience better.
+
+        if (bOpenLedger) // Ledger is not final, can vote no.
+        {
+            // Hope for more reserve to come in or more offers to consume. If we
+            // specified a local error this transaction will not be retried, so
+            // specify a tec to distribute the transaction and allow it to be
+            // retried. In particular, it may have been successful to a
+            // degree (partially filled) and if it hasn't, it might succeed.
+            terResult = tecINSUF_RESERVE_OFFER;
+        }
+        else if (!crossed)
+        {
+            // Ledger is final, insufficent reserve to create offer, processed
+            // nothing.
+            terResult = tecINSUF_RESERVE_OFFER;
+        }
+        else
+        {
+            // Ledger is final, insufficent reserve to create offer, processed
```

Comment:
Run on sentence!  Also, isn't the key point that we now charge a fee?

(Why use the word final rather than our usual word, closed?)

What about something like:  `There are insufficient reserves to create the offer, but we did some processing, so we have to charge a fee.`?